apollo-router/src/plugins/fleet_detector.rs

```diff
@@ -0,0 +1,213 @@
+use std::env::consts::ARCH;
+use std::sync::atomic;
+use std::sync::atomic::AtomicU64;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::time::Instant;
+
+use opentelemetry::metrics::MeterProvider;
+use opentelemetry_api::metrics::ObservableGauge;
+use opentelemetry_api::metrics::Unit;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use sysinfo::System;
+use tokio::task::JoinHandle;
+use tower::BoxError;
+use tracing::debug;
+
+use crate::metrics::meter_provider;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+
+const REFRESH_INTERVAL: Duration = Duration::from_secs(60);
+
+#[derive(Debug, Default, Deserialize, JsonSchema)]
+struct Conf {}
+
+#[derive(Debug)]
+struct SystemGetter {
+    system: System,
+    start: Instant,
+    // Duration in milliseconds since the start time, used because Instant is not atomic
+    // and multiple threads may try to access the system
+    refresh_at: AtomicU64,
+}
+
+impl SystemGetter {
+    fn new() -> Self {
+        let mut system = System::new();
+        system.refresh_all();
+        Self {
+            system,
+            start: Instant::now(),
+            refresh_at: AtomicU64::new(REFRESH_INTERVAL.as_millis() as u64),
+        }
+    }
+
+    fn get_system(&mut self) -> &System {
+        let refresh_at = Duration::from_millis(self.refresh_at.load(atomic::Ordering::Relaxed));
+        if self.start.elapsed() < refresh_at {
+            &self.system
+        } else {
+            self.system.refresh_cpu_all();
+            self.system.refresh_memory();
+            self.refresh_at = ((refresh_at + REFRESH_INTERVAL).as_millis() as u64).into();
+            &self.system
+        }
+    }
+}
+
+#[derive(Debug)]
+struct FleetDetector {
+    handle: JoinHandle<()>,
+    #[allow(dead_code)]
+    // We have to store a reference to the gauge otherwise it will be dropped once the plugin is
+    // initialised, even though it still has data to emit
+    freq_gauge: ObservableGauge<u64>,
```

Comment:
@BrynCooke Could you confirm it will handle the hot reload properly ?

Reply:
Yes I believe it will, but do test